### PR TITLE
Solar zenith angle calculation

### PR DESCRIPTION
**Pull request by milankl**
Following [Di Napoli, 2020](https://link.springer.com/article/10.1007/s00484-020-01900-5)

image

### CODE REVIEW

**Comment by milankl** [comment]
Illustration of the time correction, also alled [equation of time](https://en.wikipedia.org/wiki/Equation_of_time), quantifying how sunrise/sunset have an additional oscillation of up to +-16min throughout the year

image

Maybe unnecessarily precise for now ... but hey.

**Comment by milankl** [comment]
No daily cycle but seasonal cycle, Jan 1

image

Summer solstice 
image

With daily cycle Jan 1, midnight UTC

image

Summer solstice, 12:00 UTC

image

**Comment by milankl** [comment]
image

File: src/dynamics/spectral_grid.jl [hunk 1]
```diff
@@ -121,18 +121,27 @@ Base.@kwdef struct Geometry{NF<:AbstractFloat} <: AbstractGeometry{NF}     # NF:
...
+    "longitude (0˚...360˚) for each grid point in ring order"
     londs::Vector{NF} = get_latdlonds(Grid,nlat_half)[2]
     
     "latitude (-90˚...˚90) for each grid point in ring order"
     latds::Vector{NF} = get_latdlonds(Grid,nlat_half)[1]
 
+    "longitude (0...2π) for each grid point in ring order"
+    lons::Vector{NF} = RingGrids.get_latlons(Grid,nlat_half)[2]
+    
+    "latitude (-π...π) for each grid point in ring order"
```
Comment (simone-silvestri) [inline src/dynamics/spectral_grid.jl:142]:
```suggestion
    "latitude (-π/2...π/2) for each grid point in ring order"
```
😄

File: src/physics/zenith.jl [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    println(io,"$(typeof(L)) <: AbstractSolarTimeCorrection")
+    keys = propertynames(L)
+    print_fields(io,L,keys)
+end
+
+"""
+$(TYPEDSIGNATURES)
+Chooses from SolarZenith (daily and seasonal cycle), SolarZenithDay (daily cycle only),
+SolarZenithSeason (seasonal cycle only) and SolarZenithConstant (no daily or seasonal cycle)
+given the parameters in model.planet."""
```
Comment (simone-silvestri) [inline src/physics/zenith.jl:105]: Solar Zenith constant would be the latitudinal, yearly averaged, incoming solar radiation?

File: src/physics/zenith.jl [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+    rings = eachring(cos_zenith)
+    @inbounds for (j,ring) in enumerate(rings)
+            
+        ϕ = lat[j]
+        h₀ = abs(δ) + abs(ϕ) < π/2 ?        # polar day/night?
+        acos(-tan(ϕ) * tan(δ)) :            # if not: calculate length of day
+        ϕ*δ > 0 ? π : 0                     # polar day if signs are equal, otherwise polar night
+        
+        sinϕ, cosϕ = sinlat[j], coslat[j]
+        cos_zenith_j = max(0,h₀*sinδ*sinϕ + cosδ*cosϕ*sin(h₀))
```
Comment (simone-silvestri) [inline src/physics/zenith.jl:276]:
```suggestion
        cos_zenith_j = h₀*sinδ*sinϕ + cosδ*cosϕ*sin(h₀)
```
probably don't need a max here

**Comment by milankl** [comment]
Seasonal cycle is symmetric for `::SolarZenith` 👍🏼 

image

**Comment by milankl** [comment]
And for `::SolarZenithSeason` 👍🏼 

image

**Pull request merged**